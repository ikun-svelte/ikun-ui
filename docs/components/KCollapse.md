---
title: KCollapse
lang: en-US
---

# KCollapse

Use Collapse to store contents.

## Install

::: code-group

```bash [pnpm]
pnpm add @ikun-ui/collapse
```

```bash [yarn]
yarn add @ikun-ui/collapse
```

```bash [npm]
npm install @ikun-ui/collapse
```

:::

## Basic usage

You can expand multiple panels

<demo src="collapse/basic.svelte" github="Collapse"></demo>

## Custom title and content

You can render content through slots

<demo src="collapse/custom.svelte" github="Collapse"></demo>

## Accordion

Use the `KCollapseWrapper` component to wrap and set the `uid` to enable the accordion style

<demo src="collapse/accordion.svelte" github="Collapse"></demo>

## Collapse Props

| Name      | Type                     | Default | Description                          |
| --------- | ------------------------ | ------- | ------------------------------------ |
| title     | `string`                 | `-`     | Content of Collapse Item title       |
| content   | `string`                 | `-`     | Content of Collapse Item             |
| show      | `boolean`                | `false` | Expand the Collapse                  |
| showClose | `boolean`                | `true`  | Determine whether display close ionc |
| uid       | `string`                 | `-`     | Incoming in accordion style          |
| cls       | `string`                 | `-`     | Additional class                     |
| attrs     | `Record<string, string>` | `{}`    | Additional attributes                |

## Collapse Events

| Name   | Description                            | Type                   |
| ------ | -------------------------------------- | ---------------------- |
| change | Event fired when the `show` is changes | `(show: Event)=> void` |

## Collapse Slots

| Name      | Description                           |
| --------- | ------------------------------------- |
| title     | Customize Collapse Item title content |
| content   | Customize Collapse Item content       |
| closeIcon | Customize close icon                  |

## Collapse Wrapper Props

| Name      | Type                     | Default | Description                    |
| --------- | ------------------------ | ------- | ------------------------------ |
| accordion | `boolean`                | `false` | Whether to use accordion style |
| cls       | `string`                 | `-`     | Additional class               |
| attrs     | `Record<string, string>` | `{}`    | Additional attributes          |

## Collapse Wrapper Slots

| Name    | Description                        |
| ------- | ---------------------------------- |
| default | Customize Collapse wrapper content |
