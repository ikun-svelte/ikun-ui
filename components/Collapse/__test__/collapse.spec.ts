import { tick } from 'svelte';
import { afterEach, expect, test, describe, beforeEach, vi } from 'vitest';
import KCollapse from '../src/index.svelte';
import KCollapseContent from './fixture/collapse.content.test.svelte';
import KCollapseTitle from './fixture/collapse.title.test.svelte';
import KCollapseCloseIcon from './fixture/collapse.close.test.svelte';
let host: HTMLElement;

const initHost = () => {
	host = document.createElement('div');
	host.setAttribute('id', 'host');
	document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
	vi.useFakeTimers();
});
afterEach(() => {
	host.remove();
	vi.restoreAllMocks();
});

describe('Test: KCollapse', () => {
	test('props: title', async () => {
		const instance = new KCollapse({
			target: host,
			props: {
				title: 'title-content'
			}
		});
		expect(instance).toBeTruthy();
		const titleEl = host.children[0].children[0];
		expect(titleEl.getAttribute('class')).toBe('k-collapse--title k-collapse--title__dark');
		expect(titleEl.textContent).toBe('title-content ');
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: showClose', async () => {
		const instance = new KCollapse({
			target: host,
			props: {
				title: 'title-content',
				showClose: false
			}
		});
		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('k-icon--bas')).not.toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: content', async () => {
		const instance = new KCollapse({
			target: host,
			props: {
				show: true,
				content: 'foo'
			}
		});
		expect(instance).toBeTruthy();
		const contentEl = host.children[0].children[1];
		expect(contentEl.textContent).toBe(' foo');
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: show', async () => {
		const instance = new KCollapse({
			target: host,
			props: {
				show: false,
				content: 'foo'
			}
		});
		expect(instance).toBeTruthy();
		instance.$set({ show: true });
		await tick();
		expect(host.children[0].children.length).toBe(2);
		expect(host.innerHTML).matchSnapshot();
	});

	test('event: change', async () => {
		const mockFn = vi.fn();
		let show = false;
		const instance = new KCollapse({
			target: host,
			props: {
				show,
				content: 'foo'
			}
		});
		await tick();
		instance.$on('change', (e) => {
			show = e.detail;
			mockFn();
		});
		expect(instance).toBeTruthy();
		expect(host.children[0].children.length).toBe(1);
		const triggerEl = host.children[0].children[0].children[0];
		triggerEl.dispatchEvent(new window.Event('click', { bubbles: true }));
		await tick();
		await vi.advanceTimersByTimeAsync(300);
		expect(host.children[0].children.length).toBe(2);
		expect(mockFn).toBeCalledTimes(1);
		expect(show).toBeTruthy();
		triggerEl.dispatchEvent(new window.Event('click', { bubbles: true }));
		instance.$set({ show });
		await tick();
		await vi.advanceTimersByTimeAsync(300);
		expect(mockFn).toBeCalledTimes(2);
		expect(show).not.toBeTruthy();
	});

	test('slot: title', async () => {
		const instance = new KCollapseTitle({
			target: host
		});
		expect(instance).toBeTruthy();
		expect(document.getElementById('custom_title')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('slot: content', async () => {
		const instance = new KCollapseContent({
			target: host
		});
		expect(instance).toBeTruthy();
		expect(document.getElementById('custom_content')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('slot: closeIcon', async () => {
		const instance = new KCollapseCloseIcon({
			target: host
		});
		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('üçì closeIcon')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});
});
