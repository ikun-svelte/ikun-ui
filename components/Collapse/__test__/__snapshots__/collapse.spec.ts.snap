// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KCollapse > props: content 1`] = `"<div class=\\"k-collapse--base k-collapse--base__dark \\"><div class=\\"k-collapse--title k-collapse--title__dark\\" aria-hidden=\\"true\\"> <div role=\\"\\" aria-hidden=\\"true\\" class=\\"k-icon--base k-icon--base__dark  \\"><div class=\\"i-carbon-chevron-right rotate-90 k-icon-transition \\" style=\\"width: 24px; height: 24px;\\"></div></div></div> <div class=\\"k-collapse--content\\"><div class=\\"k-collapse--line\\"></div> foo</div></div>"`;

exports[`Test: KCollapse > props: show 1`] = `"<div class=\\"k-collapse--base k-collapse--base__dark \\"><div class=\\"k-collapse--title k-collapse--title__dark\\" aria-hidden=\\"true\\"> <div role=\\"\\" aria-hidden=\\"true\\" class=\\"k-icon--base k-icon--base__dark  \\"><div class=\\"i-carbon-chevron-right rotate-90 k-icon-transition \\" style=\\"width: 24px; height: 24px;\\"></div></div></div> <div class=\\"k-collapse--content\\" style=\\"animation: __svelte_516695038_0 300ms linear 0ms 1 both;\\"><div class=\\"k-collapse--line\\"></div> foo</div></div>"`;

exports[`Test: KCollapse > props: title 1`] = `"<div class=\\"k-collapse--base k-collapse--base__dark \\"><div class=\\"k-collapse--title k-collapse--title__dark\\" aria-hidden=\\"true\\">title-content <div role=\\"\\" aria-hidden=\\"true\\" class=\\"k-icon--base k-icon--base__dark  \\"><div class=\\"i-carbon-chevron-right  k-icon-transition \\" style=\\"width: 24px; height: 24px;\\"></div></div></div> </div>"`;

exports[`Test: KCollapse > slot: content 1`] = `"<div class=\\"k-collapse--base k-collapse--base__dark \\"><div class=\\"k-collapse--title k-collapse--title__dark\\" aria-hidden=\\"true\\"> <div role=\\"\\" aria-hidden=\\"true\\" class=\\"k-icon--base k-icon--base__dark  \\"><div class=\\"i-carbon-chevron-right rotate-90 k-icon-transition \\" style=\\"width: 24px; height: 24px;\\"></div></div></div> <div class=\\"k-collapse--content\\"><div class=\\"k-collapse--line\\"></div> <div slot=\\"content\\" class=\\"flex\\"><div id=\\"custom_content\\">🍞 Bread! Bread! We want to bread!</div></div></div></div>"`;

exports[`Test: KCollapse > slot: title 1`] = `"<div class=\\"k-collapse--base k-collapse--base__dark \\"><div class=\\"k-collapse--title k-collapse--title__dark\\" aria-hidden=\\"true\\"><div slot=\\"title\\" class=\\"flex\\"><div id=\\"custom_title\\">🍓 Germinal</div></div> <div role=\\"\\" aria-hidden=\\"true\\" class=\\"k-icon--base k-icon--base__dark  \\"><div class=\\"i-carbon-chevron-right  k-icon-transition \\" style=\\"width: 24px; height: 24px;\\"></div></div></div> </div>"`;
